Nama  : Michelle Sianturi
NIM   : 42324062
Prodi : D3 Teknologi Informasi

In [1]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_iris
# Load dataset Iris dari scikit-learn
iris = load_iris()

Kode ini adalah langkah persiapan: memanggil alat-alat yang dibutuhkan, lalu mengambil dataset yang akan dipakai.

Tiga baris pertama adalah proses import, yaitu memanggil library agar bisa dipakai di program. import numpy as np memanggil NumPy, library untuk mengolah angka dan array, dan as np memberi nama singkat np supaya penulisannya tidak panjang. import pandas as pd memanggil pandas, library untuk mengolah data berbentuk tabel (DataFrame), dengan nama singkat pd. Baris from sklearn.datasets import load_iris mengambil satu fungsi saja dari scikit-learn, yaitu load_iris, yang tugasnya memuat dataset Iris. Dataset ini sudah tersedia di dalam scikit-learn, jadi kamu tidak perlu mengunduh file apa pun.

Baris yang diawali tanda # adalah komentar, yaitu catatan untuk manusia yang membaca kode. Python mengabaikannya saat program berjalan.

Baris terakhir, iris = load_iris(), menjalankan fungsi tadi dan menyimpan hasilnya di variabel iris. Isinya bukan tabel biasa, melainkan objek yang menyimpan beberapa bagian sekaligus: iris.data berisi angka ukuran bunga (150 baris, 4 kolom), iris.target berisi label spesies dalam bentuk angka 0, 1, dan 2, iris.feature_names berisi nama keempat kolom fitur, dan iris.target_names berisi nama spesiesnya (setosa, versicolor, virginica). Bagian-bagian inilah yang dipakai di langkah berikutnya untuk membuat DataFrame.

In [2]:
# Konversi data menjadi DataFrame Pandas
df = pd.DataFrame(iris.data, columns=iris.feature_names)
df["Species"] = iris.target_names[iris.target]
# Definisikan X (fitur) dan y (target/label)
X = df.drop(columns=["Species"])
y = df["Species"]
print(X)

     sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                  5.1               3.5                1.4               0.2
1                  4.9               3.0                1.4               0.2
2                  4.7               3.2                1.3               0.2
3                  4.6               3.1                1.5               0.2
4                  5.0               3.6                1.4               0.2
..                 ...               ...                ...               ...
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

[150 rows x 4 columns]


Kode ini mengubah data mentah Iris menjadi tabel yang rapi, lalu memisahkannya menjadi bagian fitur dan bagian target.

Baris df = pd.DataFrame(iris.data, columns=iris.feature_names) membuat tabel pandas (DataFrame) bernama df. Isi tabelnya diambil dari iris.data, yaitu 150 baris angka ukuran bunga dengan 4 kolom. Parameter columns=iris.feature_names memberi nama pada keempat kolom itu: sepal length, sepal width, petal length, dan petal width (semuanya dalam cm). Tanpa parameter ini, kolomnya hanya bernomor 0 sampai 3.

Baris berikutnya, df["Species"] = iris.target_names[iris.target], menambahkan kolom baru bernama Species. Di sini iris.target berisi label angka (0, 1, 2), sedangkan iris.target_names berisi nama spesiesnya (setosa, versicolor, virginica). Dengan menuliskan iris.target_names[iris.target], setiap angka diganti dengan nama spesies yang sesuai, misalnya angka 0 menjadi "setosa". Jadi setiap baris di tabel sekarang punya ukuran bunga sekaligus nama spesiesnya.

Setelah itu, data dipisah menjadi dua bagian sesuai kebutuhan machine learning. X = df.drop(columns=["Species"]) mengambil semua kolom kecuali Species, sehingga X berisi empat fitur ukuran bunga. Ini adalah informasi yang dipakai model untuk menebak. Lalu y = df["Species"] hanya mengambil kolom Species, yaitu jawaban yang benar (target atau label) yang ingin ditebak oleh model. Singkatnya, X adalah soalnya dan y adalah kunci jawabannya.

Terakhir, print(X) menampilkan isi X di layar. Kamu akan melihat tabel dengan 4 kolom dan keterangan [150 rows x 4 columns] di bagian bawah. Karena datanya banyak, pandas hanya menampilkan beberapa baris awal dan akhir, dan sisanya diringkas dengan tanda titik-titik.

In [3]:
y

0         setosa
1         setosa
2         setosa
3         setosa
4         setosa
         ...    
145    virginica
146    virginica
147    virginica
148    virginica
149    virginica
Name: Species, Length: 150, dtype: str

Kode y yang hanya berisi satu huruf ini artinya "tampilkan isi variabel y". Di notebook (Jupyter atau VS Code), kalau baris terakhir dalam sebuah sel hanya berisi nama variabel, isinya otomatis ditampilkan sebagai output, tanpa perlu print(). Jadi kode ini tidak mengubah apa pun, hanya dipakai untuk mengecek isi y.

Di modul, y muncul dua kali dan outputnya berbeda karena isinya sudah berubah. Yang pertama muncul sebelum LabelEncoder, saat y masih berupa kolom Species dari DataFrame. Outputnya berupa daftar bernomor dengan nama spesies (setosa, versicolor, virginica), diakhiri keterangan Name: Species, Length: 150, dtype: object. Artinya ada 150 data, bernama Species, dan bertipe teks (object). Bagian angka di kiri (0, 1, 2, dan seterusnya) hanyalah nomor baris.

Yang kedua muncul setelah y = lab.fit_transform(y). Saat itu nama spesies sudah diubah menjadi angka, sehingga outputnya berupa array seperti array([0, 0, 0, ..., 1, 1, 1, ..., 2, 2, 2]). Angka 0 mewakili setosa, 1 versicolor, dan 2 virginica (urutannya alfabetis). Karena datanya berurutan per spesies, kamu melihat 50 angka 0, lalu 50 angka 1, lalu 50 angka 2. Alasan diubah ke angka adalah model machine learning bekerja dengan angka, bukan teks.

In [4]:
from sklearn.preprocessing import LabelEncoder
lab = LabelEncoder()
y = lab.fit_transform(y)
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

Kode ini mengubah label spesies yang berupa teks menjadi angka, karena model machine learning bekerja dengan angka, bukan kata.

Baris from sklearn.preprocessing import LabelEncoder memanggil alat bernama LabelEncoder dari scikit-learn. Tugasnya memberi nomor pada setiap kategori teks. Baris lab = LabelEncoder() lalu membuat satu objek dari alat itu dan menyimpannya di variabel lab. Anggap saja ini seperti menyiapkan mesin penerjemah yang masih kosong.

Baris y = lab.fit_transform(y) menjalankan dua langkah sekaligus. Bagian fit membuat mesin itu mempelajari kategori apa saja yang ada di y, yaitu setosa, versicolor, dan virginica, lalu mengurutkannya secara alfabetis dan memberi nomor: setosa = 0, versicolor = 1, virginica = 2. Bagian transform kemudian mengganti setiap nama spesies di y dengan nomornya. Hasilnya disimpan kembali ke variabel y, sehingga isi y yang lama (teks) tertimpa oleh versi angkanya.

Baris terakhir, y, hanya menampilkan isinya di layar. Outputnya berupa array berisi 150 angka, yaitu 50 angka 0, lalu 50 angka 1, lalu 50 angka 2, sesuai urutan data aslinya.

Objek lab sengaja disimpan karena masih dibutuhkan nanti. Saat model memberi hasil prediksi berupa angka (misalnya 2), kamu memakai lab.inverse_transform(result) untuk mengembalikannya menjadi nama spesies (virginica). Itulah yang terjadi di bagian prediksi data baru.

In [5]:
from sklearn.model_selection import train_test_split
X_train, X_test = train_test_split(X, test_size=0.2, random_state=0)
y_train, y_test = train_test_split(y, test_size=0.2, random_state=0)

Kode ini membagi data menjadi dua kelompok: data latih untuk mengajari model, dan data uji untuk menguji seberapa pintar model setelah belajar. Kalau model diuji dengan data yang sama seperti yang dipakai belajar, hasilnya tidak jujur, seperti murid yang diberi soal ujian yang sudah pernah dibahas.

Baris from sklearn.model_selection import train_test_split memanggil fungsi pembagi data dari scikit-learn. Fungsi ini mengacak urutan data terlebih dulu, lalu memotongnya menjadi dua bagian sesuai proporsi yang kita tentukan.

Baris X_train, X_test = train_test_split(X, test_size=0.2, random_state=0) membagi fitur (X). Parameter test_size=0.2 artinya 20% data dijadikan data uji, sehingga sisanya, 80%, menjadi data latih. Dari 150 data, hasilnya 120 untuk latih (X_train) dan 30 untuk uji (X_test). Parameter random_state=0 mengunci pengacakan supaya hasil pembagiannya selalu sama setiap kali kode dijalankan. Tanpa itu, pembagiannya berubah-ubah, sehingga akurasinya pun bisa berbeda tiap kali dijalankan.

Baris berikutnya, y_train, y_test = train_test_split(y, test_size=0.2, random_state=0), melakukan hal yang sama untuk label (y), yaitu kunci jawabannya. Karena test_size dan random_state yang dipakai sama persis, pengacakan yang terjadi juga identik. Akibatnya, baris ke-1 di X_train tetap berpasangan dengan baris ke-1 di y_train, dan begitu juga untuk data uji. Kalau random_state-nya berbeda, soal dan kunci jawabannya bisa tertukar dan model akan belajar dengan data yang salah pasangan.

Setelah kode ini, akan ada empat variabel: X_train dan y_train untuk melatih model, serta X_test dan y_test untuk menguji model. Cara yang lebih ringkas sebenarnya adalah X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0), tapi versi di modul juga menghasilkan pembagian yang sama.

In [6]:
from sklearn.naive_bayes import GaussianNB
model = GaussianNB()
NB_model = model.fit(X_train, y_train)

Kode ini membuat model Naïve Bayes lalu melatihnya memakai data latih, sehingga model belajar mengenali pola ukuran bunga untuk tiap spesies.

Baris from sklearn.naive_bayes import GaussianNB memanggil algoritma Gaussian Naïve Bayes dari scikit-learn. Pilihan "Gaussian" dipakai karena keempat fitur Iris berupa angka kontinu (panjang dan lebar dalam cm) yang diasumsikan berdistribusi normal.

Baris model = GaussianNB() membuat sebuah model kosong dan menyimpannya di variabel model. Pada tahap ini model belum tahu apa-apa, ibarat murid baru yang belum belajar.

Baris NB_model = model.fit(X_train, y_train) adalah proses belajarnya. Fungsi fit() menerima X_train (ukuran bunga) dan y_train (spesiesnya), lalu untuk setiap spesies model menghitung rata-rata dan standar deviasi dari masing-masing fitur, serta seberapa sering spesies itu muncul di data latih. Angka-angka inilah yang nanti dipakai dalam rumus Gaussian dan Teorema Bayes saat model menebak data baru. Misalnya, model mencatat bahwa petal width setosa rata-ratanya sangat kecil, sedangkan virginica jauh lebih besar.

Hasil dari fit() disimpan ke NB_model. Sebenarnya fit() mengembalikan objek model yang sama dengan model, jadi NB_model dan model mengarah ke model terlatih yang sama. Itu sebabnya nanti kamu bisa memanggil NB_model.predict(...) untuk menebak data uji maupun data baru.

In [7]:
result = NB_model.predict(X_test)
result

array([2, 1, 0, 2, 0, 2, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 2, 1,
       0, 0, 2, 0, 0, 1, 1, 0])

Kode ini meminta model yang sudah dilatih untuk menebak spesies dari data uji, yaitu data yang belum pernah dilihat model saat belajar.

Baris result = NB_model.predict(X_test) memberikan X_test (30 baris ukuran bunga) ke model. Untuk setiap baris, model menghitung probabilitas bunga itu termasuk setosa, versicolor, atau virginica, memakai rata-rata, standar deviasi, dan frekuensi kelas yang dipelajari saat fit(). Kelas dengan probabilitas tertinggi dipilih sebagai tebakan. Hasilnya disimpan di variabel result. Perhatikan bahwa y_test (kunci jawaban) tidak diberikan ke model di tahap ini, karena model harus menebak sendiri.

Baris terakhir, result, hanya menampilkan isinya di layar. Outputnya berupa array berisi 30 angka, misalnya array([2, 1, 0, 2, 0, ...]). Angka-angka itu adalah tebakan model dalam bentuk label numerik: 0 untuk setosa, 1 untuk versicolor, dan 2 untuk virginica. Urutannya mengikuti urutan baris di X_test, jadi angka pertama adalah tebakan untuk baris pertama data uji, dan seterusnya.

In [8]:
from sklearn.metrics import accuracy_score
print('Accuracy of data testing: ', accuracy_score(y_test, result)*100)

Accuracy of data testing:  96.66666666666667


Kode ini mengukur seberapa akurat tebakan model, yaitu dengan membandingkan hasil tebakan dengan kunci jawaban yang sebenarnya.

Baris from sklearn.metrics import accuracy_score memanggil fungsi penghitung akurasi dari scikit-learn. Akurasi adalah persentase tebakan yang benar dari seluruh data uji.

Di dalam print, fungsi accuracy_score(y_test, result) menerima dua daftar. y_test adalah jawaban asli untuk 30 data uji, sedangkan result adalah tebakan model untuk data yang sama. Fungsi ini membandingkannya satu per satu berdasarkan urutan, menghitung berapa yang cocok, lalu membaginya dengan jumlah seluruh data. Hasilnya berupa angka antara 0 dan 1. Misalnya, kalau 29 dari 30 tebakan benar, hasilnya 29/30 = 0,9667.

Bagian *100 mengubah angka itu menjadi persen, sehingga 0,9667 menjadi 96,67. Teks 'Accuracy of data testing: ' hanya label agar output mudah dibaca. Hasil yang muncul di modul adalah Accuracy of data testing: 96.66666666666667, artinya model benar menebak 29 dari 30 data uji dan hanya salah 1.

Urutan argumen di accuracy_score sebaiknya selalu kunci jawaban dulu (y_test), baru tebakan (result). Untuk akurasi biasa, kalau tertukar hasilnya sebenarnya sama, tetapi kebiasaan ini penting karena untuk metrik lain seperti precision dan recall, urutan yang tertukar bisa menghasilkan nilai yang salah.

In [9]:
#prediksi inputan data baru
#input_data = pd.DataFrame(np.array([[5.1, 3.5, 1.4, 0.2]]), columns=iris.feature_names)
input_data = [[5.1, 3.5, 1.4, 0.2]]
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: ['setosa']


c:\Users\Admin\anaconda3\Lib\site-packages\sklearn\utils\validation.py:2827: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


Kode ini dipakai untuk menebak spesies dari satu bunga baru yang ukurannya kamu tentukan sendiri, bukan dari data uji.

Baris yang diawali tanda # adalah komentar, jadi tidak dijalankan. Baris kedua yang dikomentari adalah cara lain untuk membuat input, yaitu dengan DataFrame lengkap dengan nama kolom. Cara itu baru dipakai di sel berikutnya dalam modul.

Baris input_data = [[5.1, 3.5, 1.4, 0.2]] membuat data bunga baru berupa list di dalam list. Empat angkanya adalah sepal length, sepal width, petal length, dan petal width, sesuai urutan kolom saat model dilatih. Tanda kurung ganda dibutuhkan karena predict() selalu mengharapkan banyak baris sekaligus (bentuk 2 dimensi), sehingga satu bunga tetap harus dibungkus sebagai satu baris di dalam daftar.

Baris result = NB_model.predict(input_data) meminta model menebak. Hasilnya berupa angka label, misalnya [0], karena model dilatih dengan label numerik. Angka itu belum bermakna bagi manusia, sehingga baris prediction_label = lab.inverse_transform(result) menerjemahkannya kembali menjadi nama spesies memakai lab (LabelEncoder) yang dibuat sebelumnya. Angka 0 menjadi 'setosa', 1 menjadi 'versicolor', dan 2 menjadi 'virginica'. Baris print(...) lalu menampilkan hasilnya, yaitu ['setosa'].

Saat dijalankan, akan muncul UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names. Penyebabnya, model dilatih dengan data yang kolomnya punya nama, sedangkan input_data di sini hanya list biasa tanpa nama kolom. Ini hanya peringatan, bukan error, dan hasil prediksinya tetap benar. Peringatan itulah yang dihilangkan di sel berikutnya dengan membungkus input menjadi DataFrame yang memakai columns=iris.feature_names.

In [10]:
input_data = pd.DataFrame(np.array([[5.1, 3.5, 1.4, 0.2]]), columns=iris.feature_names)
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: ['setosa']


Kode ini melakukan hal yang sama dengan sel sebelumnya, yaitu menebak spesies dari satu bunga baru. Bedanya, inputnya dibungkus dalam DataFrame yang punya nama kolom, sehingga peringatan (warning) tidak muncul lagi.

Baris input_data = pd.DataFrame(np.array([[5.1, 3.5, 1.4, 0.2]]), columns=iris.feature_names) dikerjakan dari dalam ke luar. Bagian np.array([[5.1, 3.5, 1.4, 0.2]]) membuat array NumPy berisi satu baris dengan empat angka ukuran bunga, yaitu sepal length, sepal width, petal length, dan petal width. Array itu kemudian diberikan ke pd.DataFrame(...) untuk diubah menjadi tabel pandas. Parameter columns=iris.feature_names memberi nama pada keempat kolomnya, persis sama dengan nama kolom yang dipakai saat model dilatih. Jadi hasilnya tabel kecil dengan satu baris dan empat kolom bernama.

Penyebab warning di sel sebelumnya adalah model dilatih dengan data yang kolomnya bernama, sedangkan input berupa list biasa tanpa nama. Scikit-learn mengingat nama kolom saat fit(), lalu membandingkannya saat predict(). Dengan DataFrame yang namanya identik, strukturnya konsisten dengan data latih, sehingga warning hilang. Hasil prediksinya sendiri tidak berubah, hanya cara memberikan inputnya yang lebih rapi.

Tiga baris berikutnya sama seperti sebelumnya. NB_model.predict(input_data) menghasilkan tebakan berupa angka label, lab.inverse_transform(result) menerjemahkan angka itu menjadi nama spesies, dan print(...) menampilkannya. Outputnya ['setosa'], karena ukuran bunga ini (terutama petal length 1.4 dan petal width 0.2) memang sangat khas setosa.

Cara ini juga yang akan dipakai untuk tugas nomor 2. Hanya dengan mengganti angkanya menjadi [[7.3, 3.8, 4.6, 2.3]], dan strukturnya tetap sama.

Tugas

In [11]:
input_data = pd.DataFrame(np.array([[7.3, 3.8, 4.6, 2.3]]), columns=iris.feature_names)
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)
print(NB_model.predict_proba(input_data))

Hasil prediksi untuk input baru adalah: ['virginica']
[[4.83048542e-194 3.68081469e-007 9.99999632e-001]]


Kode ini adalah jawaban tugas nomor 2, yaitu menebak spesies dari bunga baru dengan ukuran sepal length 7.3, sepal width 3.8, petal length 4.6, dan petal width 2.3 (semuanya dalam cm). Strukturnya sama dengan sel contoh di modul, hanya angkanya yang diganti.

Baris pertama membuat data bunga baru. np.array([[7.3, 3.8, 4.6, 2.3]]) menyusun keempat angka itu menjadi satu baris, lalu pd.DataFrame(..., columns=iris.feature_names) mengubahnya menjadi tabel kecil dengan nama kolom yang sama seperti data latih. Dengan begitu, model bisa mengenali angka mana yang merupakan sepal length, sepal width, petal length, dan petal width, dan warning tidak muncul.

Baris kedua, result = NB_model.predict(input_data), meminta model yang sudah dilatih untuk menebak. Hasilnya berupa angka label, yaitu [2]. Baris ketiga, lab.inverse_transform(result), menerjemahkan angka itu kembali ke nama spesies memakai LabelEncoder (0 = setosa, 1 = versicolor, 2 = virginica), sehingga angka 2 menjadi virginica. Baris print menampilkannya sebagai ['virginica'].

Baris terakhir, print(NB_model.predict_proba(input_data)), menampilkan seberapa yakin model terhadap tebakannya. Outputnya tiga angka, berurutan untuk setosa, versicolor, dan virginica, dan jumlahnya 1. Hasilnya sekitar [4.8e-194, 3.7e-07, 0.9999996]. Artinya peluang setosa praktis nol, peluang versicolor sangat kecil (0,00004%), dan peluang virginica hampir 100%. Jadi model bukan sekadar menebak, tetapi sangat yakin bahwa bunga ini adalah virginica.

Kesimpulannya, dari ukuran [7.3, 3.8, 4.6, 2.3], model Gaussian Naïve Bayes memprediksi bunga tersebut sebagai virginica. Faktor yang paling berpengaruh adalah petal width 2.3 cm yang besar, karena ukuran sebesar itu jarang dimiliki setosa maupun versicolor.